core/vm/interpreter.go

```diff
@@ -826,95 +786,189 @@ func (in *GethEVMInterpreter) runPlain(state *InterpreterState, input []byte, re
-		// the operation
-		// Memory check needs to be done prior to evaluating the dynamic gas portion,
-		// to detect calculation overflows
-		if operation.memorySize != nil {
-			memSize, overflow := operation.memorySize(stack)
-			if overflow {
-				return nil, ErrGasUintOverflow
-			}
-			// memory is expanded in words of 32 bytes. Gas
-			// is also calculated in words.
-			if memorySize, overflow = math.SafeMul(toWordSize(memSize), 32); overflow {
-				return nil, ErrGasUintOverflow
-			}
+	return gethState.Result, gethState.Err
+}
+
+///////////////////////////////////////////////////////////////////////////////////////////////////
+// Abstracted interpreter with single step execution.
+
+// GethState represents the internal state of the interpreter.
+type GethState struct {
+	Contract *Contract // processed contract
+	Memory   *Memory   // bound memory
+	Stack    *Stack    // local stack
+	// For optimisation reason we're using uint64 as the program counter.
+	// It's theoretically possible to go above 2^64. The YP defines the PC
+	// to be uint256. Practically much less so feasible.
+	Pc          uint64 // program counter
+	Result      []byte // result of the opcode execution function
+	Err         error
+	CallContext *ScopeContext
+	Halted      bool
+
+	op   OpCode // current opcode
+	cost uint64
+
+	// copies used by tracer
+	pcCopy  uint64 // needed for the deferred Tracer
+	gasCopy uint64 // for Tracer to log gas remaining before execution
+	logged  bool   // deferred Tracer should ignore already logged steps
+}
+
+func NewGethState(contract *Contract, memory *Memory, stack *Stack, Pc uint64) *GethState {
+	return &GethState{
+		Contract: contract,
+		Memory:   memory,
+		Stack:    stack,
+		Pc:       Pc,
+		CallContext: &ScopeContext{
+			Memory:   memory,
+			Stack:    stack,
+			Contract: contract,
+		},
+	}
+}
+
+func (in *GethEVMInterpreter) StartRun(state *GethState, input []byte, readOnly bool) {
+	// Increment the call depth which is restricted to 1024
+	in.evm.Depth++
+
+	// Make sure the readOnly is only set if we aren't in readOnly yet.
+	// This also makes sure that the readOnly flag isn't removed for child calls.
+	if readOnly && !in.readOnly {
+		in.readOnly = true
+	}
+
+	// Reset the previous call's return data. It's unimportant to preserve the old buffer
+	// as every returning call will return new data anyway.
+	in.returnData = nil
+
+	state.Contract.Input = input
+}
+
+func (in *GethEVMInterpreter) StopRun(readOnly bool) {
+	in.evm.Depth--
+
+	// Make sure the readOnly is only set if we aren't in readOnly yet.
+	// This also makes sure that the readOnly flag isn't removed for child calls.
+	if readOnly && in.readOnly {
+		in.readOnly = false
+	}
+}
+
+func (in *GethEVMInterpreter) Step(state *GethState) bool {
+	// Don't execute if there is no code.
+	if len(state.Contract.Code) == 0 {
+		return false
+	}
+
+	// Don't execute if already halted.
+	if state.Halted {
+		return false
+	}
```

Comment:
These two checks are now executed for every single instruction, which adds overhead. Can they be removed?

Reply:
Not necessarily, no. The code length check can be done once in `runPlain`, and since `Halted` is public this can also be done on the Tosca CT side. So I moved the code length check to `runPlain` and removed the halted check.